app/src/main/kotlin/io/sweers/catchup/ui/fragments/SmmryFragment.kt

```diff
@@ -178,77 +163,70 @@ class SmmryFragment : InjectableBaseFragment() {
+      val response = tryRequestFromStorage() ?: fetchFromNetwork()
+      alreadyLoaded = true
+      loadingView.hide(animate = true)
+      when (response) {
+        is Success -> {
+          showSummary(response)
+        }
+        else -> {
+          val message = when (response) {
+            is InternalError -> "Smmry internal error - ${response.message}"
+            is IncorrectVariables -> "Smmry invalid input - ${response.message}"
+            is ApiRejection -> "Smmry API error - ${response.message}"
+            is SummarizationError -> "Smmry summarization error - ${response.message}"
+            UnknownErrorCode -> getString(R.string.unknown_issue)
+            else -> TODO("Give me sealed whens!")
           }
-        })
+          summary.text = message
+        }
+      }
+    }
   }
 
   fun canScrollVertically(directionInt: Int): Boolean {
     return content.canScrollVertically(directionInt)
   }
 
-  private fun tryRequestFromStorage(): Maybe<SmmryResponse> {
-    return smmryDao.getItem(id)
-        .map { moshi.adapter(SmmryResponse::class.java).fromJson(it.json)!! }
-        .onErrorComplete { exception ->
-          e(exception) { "Error loading smmry cache." }
-          true
-        }
+  private suspend fun tryRequestFromStorage() = withContext(Dispatchers.IO) {
+    smmryDao.getItem(id)?.let {
+      moshi.adapter(SmmryResponse::class.java).fromJson(it.json) ?: throw JsonDataException(
+          "Could not parse entry")
+    }
   }
 
-  private fun getRequestSingle(): Single<SmmryResponse> {
-    val summarizer: Single<SmmryResponse> = when (info.type) {
-      TEXT -> smmryService.summarizeText(SmmryRequestBuilder.forText()
-          .withBreak(true)
-          .keywordCount(5)
-          .sentenceCount(5)
-          .build(),
-          info.value)
-      URL -> smmryService.summarizeUrl(SmmryRequestBuilder.forUrl(info.value)
-          .withBreak(true)
-          .keywordCount(5)
-          .sentenceCount(5)
-          .build())
-      NONE -> Single.just(Success.just(inputTitle, info.value))
+  private suspend fun fetchFromNetwork(): SmmryResponse {
+    val response: SmmryResponse = try {
+      when (info.type) {
+        TEXT -> smmryService.summarizeText(SmmryRequestBuilder.forText()
+            .withBreak(true)
+            .keywordCount(5)
+            .sentenceCount(5)
+            .build(),
+            info.value)
+            .await()
+        URL -> smmryService.summarizeUrl(SmmryRequestBuilder.forUrl(info.value)
+            .withBreak(true)
+            .keywordCount(5)
+            .sentenceCount(5)
+            .build())
+            .await()
+        NONE -> Success.just(inputTitle, info.value)
+      }
+    } catch (error: Exception) {
+      // We should indicate when something's network related
+      UnknownErrorCode
     }
-    return summarizer.doOnSuccess {
-      if (it != UnknownErrorCode) {
+    if (response != UnknownErrorCode) {
+      withContext(Dispatchers.IO) {
```

Comment:
would it be worth moving this into the try and dropping the if block altogether? If you added the suspending dao wrappers mentioned later this would drop the wrapping and indentation in here quite a bit

Reply:
You might want to handle errors from this differently though

Reply:
Not sure I follow, could you elaborate (or sample snippet?)